### PR TITLE
Possible fix for #2837

### LINKED ISSUE #2837: "RuntimeWarning: overflow encountered in scalar multiply" in axonarawio
Hi everyone, 

I'm encountering the following error/warning in a spike-extraction script I made a while ago: 
I am using the detect_peaks function on a concatenated Axona recording, which has worked fine until recently. 
I've attached the script below. It's a bit long and not very elegant, but the error happens in the 'detect peaks' section, line 95-99. The main goal of the script is to detect spikes and export them so that we can use them in a manual clustering program (SpikeSort3D), I usually run it in Spyder. 

For some reason, I now encounter the following:
```
detect peaks using locally_exclusive:  17%|#6        | 1240/7390 [00:33<02:42, 37.85it/s]
C:\Users\leemburg\AppData\Local\anaconda3\envs\SInew\lib\site-packages\neo\rawio\axonarawio.py:384: RuntimeWarning: overflow encountered in scalar multiply
  offset = i_start // 3 * (bin_dict["bytes_packet"] // 2)
detect peaks using locally_exclusive: 100%|##########| 7390/7390 [02:48<00:00, 43.92it/s] 
```
I have updated to Python 3.10 and installed the latest version of spikeinterface (within the latest anaconda) and the warning/error persists. What can I do about it? 

```
import spikeinterface.extractors as se
import spikeinterface as si
import numpy as np
from probeinterface import read_prb
from spikeinterface.sortingcomponents.peak_detection import detect_peaks
from spikeinterface.preprocessing import bandpass_filter
from scipy.io import savemat
from tqdm import tqdm


InFolder = "E:/AxonaToNLXtest/"
OutFolder = "E:/AxonaToNLXtest/pos/"
Probepath ="C:/Users/leemburg/Desktop/OEphystest/"
ChanList = 'Tetlist.txt' # text file listing good and bad channels
BaseName = 'HD263-2811_'
TetList = [2,4,6] #analyse only these tetrodes (1-based, as on drive)
numsess = 3 #number of sessions to read (e.g: numsess = 4 reads bin 01 to 04)

spike_thresh = 5 # detection threshold for spike detection is spike_thresh* signal SD
spike_sign = 'pos' #detect peaks. can be: 'neg', 'pos', 'both'

# !!! need to manually edit the .set files so that all channels get read. 

##############################################################################

tetgrouping = np.array([0,0,0,0,1,1,1,1,2,2,2,2,3,3,3,3,4,4,4,4,5,5,5,5,6,6,6,6,7,7,7,7,8,8,8,8,
                        9,9,9,9,10,10,10,10,11,11,11,11,12,12,12,12,13,13,13,13,14,14,14,14,
                        15,15,15,15,15])

# read bad channel list, set bad channels to 17
tetlist = np.loadtxt(InFolder + '/' + ChanList,
                 delimiter=",")

for tetnum in range(16):
        thistet = np.where(tetgrouping==tetnum)
        thistet = np.array(thistet)
        thesewires = tetlist[tetnum,1:5]
        badwires = np.where(thesewires==0)
        badwires = np.array(badwires)
        badchans = thistet[0][badwires]
        tetgrouping[badchans]=17



# make file list
allsess = range(numsess) #count sessions zero-based
files = list()
for f in allsess:
   thissess = allsess[f]+1 #1-based session number
   InName = InFolder+BaseName+"%02d" % thissess+".bin"
   list.append(files, InName)


# make list of SI extractor recordings
recording_list = list()
for f in allsess:
   list.append(recording_list, se.AxonaRecordingExtractor(files[f]))

# merge recording objects 
MergeRec = si.concatenate_recordings(recording_list)

# add highpass filter
MergeRec_f = bandpass_filter(MergeRec, freq_min=300.0, freq_max=6000.0, margin_ms=5.0, dtype=None)

all_chan_ids = MergeRec_f.get_channel_ids()

# select a tetrode
for tetnum in TetList:

    tet_chan_ids = all_chan_ids[np.where(tetgrouping == tetnum-1)]
    tetname = TetList[tetnum-1]

    if np.size(tet_chan_ids)>2:
        
        # 4-wire tetrode
        if np.size(tet_chan_ids) == 4:
           new_chans = [0,1,2,3]
           probename = "tet4_probe.prb"
        
        # 3-wire tetrode
        if np.size(tet_chan_ids) == 3:
            new_chans = [0,1,2]
            probename = "tet3_probe.prb"
        
    myprobe = read_prb(Probepath + "/" + probename)

    #select channels and add probe
    thistet = MergeRec_f.channel_slice(tet_chan_ids, renamed_channel_ids=new_chans)
    thistet = thistet.set_probegroup(myprobe)

    # preprocess (filter)
    thistet_f = bandpass_filter(thistet, freq_min=600, freq_max=6000)

    #detect peaks
    detectradius = 30 #tetrode channel group is 10x10um square, this radius should capture all spikes in this channelgroup
    #2 versions of function from 2 versions of spikeinterface. Yuck.
    #TetPeaks = detect_peaks(thistet_f, method='locally_exclusive', peak_sign=spike_sign, detect_threshold=spike_thresh, exclude_sweep_ms=0.1, local_radius_um=detectradius, noise_levels=None,)
    TetPeaks = detect_peaks(thistet_f, method='locally_exclusive', peak_sign=spike_sign, detect_threshold=spike_thresh, exclude_sweep_ms=0.1, radius_um=detectradius, noise_levels=None,)


    # get waveforms. must be 32 samples long, I'm setting peak at sample 8 (same as NLX)
    prepeak = 8
    postpeak = 24

    print('extracting waveforms')

    allWFs = np.zeros([32, 4, len(TetPeaks['sample_index'])], dtype='int16')
    for p in tqdm(range(len(TetPeaks['sample_index'])), desc="collecting waveforms"):
        sf = TetPeaks['sample_index'][p] - prepeak
        ef = TetPeaks['sample_index'][p] + postpeak

        thisWF = thistet_f.get_traces(segment_index=None, start_frame=sf, end_frame=ef)

        #write only complete spike waveforms (might skip last spike if too short)
        if np.size(thisWF, 0) == 32:
           allWFs[:, :, p] = thisWF

        del thisWF

   #save peaks to mat file (for later processing with Mat2NlxSpike to generate .ntt file)
    print('saving to mat file')
    outname = OutFolder+"tt"+str(tetname) + '.mat'
    savemat(outname, {"Timestamps":TetPeaks['sample_index'], "Spikes": allWFs})        
```

### PR DESCRIPTION
**Pull request by h-mayorquin**
Could fix #2837 @zm711 

The problem as has appeared before is that sometimes num_samples returns a numpy scalar which overflows.

This is solved in this PR in two ways:
1) cumsum in concatenation is determined with a numpy function, I switched to numpy list.
2) length in chunk_with_margin is forced to be an int as that is used to determine start_frame and stop_frame in a `get_traces` call.

I still need to take a look at the function in axona but this should eliminate some problems in spikeinterface. [EDIT] I looked at the function in Axona and I am convinced that the problem is in spikeinterface.

I also removed some unused imports.

### CODE REVIEW

File: src/spikeinterface/core/segmentutils.py [hunk 1]
```diff
@@ -156,7 +156,7 @@ def __init__(self, parent_segments, sampling_frequency, ignore_times=True):
         BaseRecordingSegment.__init__(self, **time_kwargs)
         self.parent_segments = parent_segments
         self.all_length = [rec_seg.get_num_samples() for rec_seg in self.parent_segments]
-        self.cumsum_length = np.cumsum([0] + self.all_length)
+        self.cumsum_length = [0] + [sum(self.all_length[: i + 1]) for i in range(len(self.all_length))]
```
Comment (samuelgarcia) [inline src/spikeinterface/core/segmentutils.py:159]: Why list instead of numpy arrays ? Is it because of the type of scalar that are not python type ?

Comment (h-mayorquin) [reply src/spikeinterface/core/segmentutils.py:159]: Yes, this is used to determine start and stop frames below.

Comment (samuelgarcia) [reply src/spikeinterface/core/segmentutils.py:159]: Yes I know but why list is better that numpy arrays for this ?

Comment (h-mayorquin) [reply src/spikeinterface/core/segmentutils.py:159]: If you get an element from a numpy array (e.g. array[index] with type int of some sort) the resulting type is not an int but a scalar array int that can overflow. This avoids having scalar arrays.

**Review by alejoe91 — APPROVED** [review]

**Pull request merged**